fieldtracing/fieldtracing.cpp

```diff
@@ -0,0 +1,1701 @@
+            MPI_Allreduce(nodeTracingCoordinates.data(), sumNodeTracingCoordinates.data(), 3*nodes.size(), MPI_FLOAT, MPI_SUM, MPI_COMM_WORLD);
+            MPI_Allreduce(nodeTracingStepSize.data(), reducedNodeTracingStepSize.data(), nodes.size(), MPI_FLOAT, MPI_MAX, MPI_COMM_WORLD);
+         }
+         for(uint n=0; n<nodes.size(); n++) {
+            if(sumNodeNeedsContinuedTracing[n] > 0) {
+               anyNodeNeedsTracing=true;
+               nodeNeedsContinuedTracing[n] = 1;
+               
+               // Update that nodes' tracing coordinates
+               nodeTracingCoordinates[n][0] = sumNodeTracingCoordinates[n][0] / sumNodeNeedsContinuedTracing[n];
+               nodeTracingCoordinates[n][1] = sumNodeTracingCoordinates[n][1] / sumNodeNeedsContinuedTracing[n];
+               nodeTracingCoordinates[n][2] = sumNodeTracingCoordinates[n][2] / sumNodeNeedsContinuedTracing[n];
+               
+               nodeStepCounter[n] = maxNodeStepCounter[n];
+            }
+            nodeTracingStepSize[n] = reducedNodeTracingStepSize[n];
+         }
+      } while(anyNodeNeedsTracing);
+      
+      logFile << "open-closed tracing traced in " << toto << " iterations of the tracing loop." << endl;
+      
+      bool redWarning = false;
+      MPI_Allreduce(&warnMaxStepsExceeded, &redWarning, 1, MPI_C_BOOL, MPI_LOR, MPI_COMM_WORLD);
+      int rank;
+      MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+      if(redWarning && rank == MASTER_RANK) {
+         logFile << "Warning: reached the maximum number of tracing steps " << maxTracingSteps << " allowed for open-closed ionosphere tracing." << endl;
+      }
+      
+      std::vector<int> reducedNodeMapping(nodes.size());
+      std::vector<int> reducedNodeTracingStepCount(nodes.size());
+      MPI_Allreduce(nodeMapping.data(), reducedNodeMapping.data(), nodes.size(), MPI_INT, MPI_MIN, MPI_COMM_WORLD);
+      MPI_Allreduce(nodeTracingStepCount.data(), reducedNodeTracingStepCount.data(), nodes.size(), MPI_INT, MPI_SUM, MPI_COMM_WORLD);
+      
+      for(uint n=0; n<nodes.size(); n++) {
+         nodes[n].openFieldLine = reducedNodeMapping.at(n);
+      }
+      
+      phiprof::stop("ionosphere-openclosedTracing");
+   }
+
+   /*! Trace magnetic field lines forward and backward from each DCCRG cell to record the connectivity.
+   */
+   void traceFullBoxConnection(
+      FsGrid< fsgrids::technical, FS_STENCIL_WIDTH> & technicalGrid,
+      FsGrid< std::array<Real, fsgrids::bfield::N_BFIELD>, FS_STENCIL_WIDTH> & perBGrid,
+      FsGrid< std::array<Real, fsgrids::dperb::N_DPERB>, FS_STENCIL_WIDTH> & dPerBGrid,
+      dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGrid
+   ) {
+      phiprof::start("ionosphere-fullTracing");
+      
+      std::vector<CellID> localDccrgCells = getLocalCells();
+      int localDccrgSize = localDccrgCells.size();
+      int globalDccrgSize;
+      MPI_Allreduce(&localDccrgSize, &globalDccrgSize, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
```

Comment:
Isn't this something that DCCRG already has locally available, as part of its global topology information?

Reply:
I found
```
        /*!
        Returns all cells in the grid that don't have children (e.g. leaf cells).

        Only those cells are returned which this process knows about, this
        might not include all cells that exist on all processes.

        By default returned cells are in random order but if sorted == true
        they are sorted using std::sort before returning.

        \see
        get_cells()
        */
        std::vector<uint64_t> get_all_cells(const bool sorted = false)
```
which doesn't guarantee these are all. Do you have another suggestion? I didn't find anything "obvious" so far looking for `cells` or `size`.